### PR TITLE
Health Check Publisher options and example

### LINKED ISSUE #10560: Add default timeout and how to change
Question:As per the documentation below, the health checks are executed periodically if we add publisher. What’s the default interval for auto execution of tests and is there any override?

Per `@pranavkm` 	

It defaults to 30 seconds and you can configure it using HealthCheckPublisherOptions.Period:

`services.Configure<HealthCheckPublisherOptions>(options => options.Period = ...);`

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c0cbb094-29e9-740a-9fd6-db372641d753
* Version Independent ID: 52d8efc9-a336-71d9-6d90-98173cbf6bf4
* Content: [Health checks in ASP.NET Core](https://docs.microsoft.com/en-in/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-2.2#health-check-publisher)
* Content Source: [aspnetcore/host-and-deploy/health-checks.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/health-checks.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #10664: Health checks at startup
Is there a way to execute all the health checks on app startup, and output an error message to the console / logger if any of them are failing? 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c0cbb094-29e9-740a-9fd6-db372641d753
* Version Independent ID: 52d8efc9-a336-71d9-6d90-98173cbf6bf4
* Content: [Health checks in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-2.2)
* Content Source: [aspnetcore/host-and-deploy/health-checks.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/health-checks.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #10560 
Fixes #10664 

Blocked by https://github.com/aspnet/Extensions/issues/1041

### CODE REVIEW

**Comment by guardrex** [comment]
@Rick-Anderson You said to ping u and Pranav, but Ryan built the HealthChecks API. When I'm done, do you still want Pranav to review? As soon as the issue is resolved with the options, I'll be able to submit this one.

**Comment by Rick-Anderson** [comment]
Fixed in #10734
Can you fetch origin?

**Comment by guardrex** [comment]
I'll get it in the morning. It's 9pm here ... Just headed to :zzz:

**Comment by guardrex** [comment]
@Rick-Anderson Looks good. I hope that's not going to be a new prob ... we have a *TON* of BOMs all over the repo. Although they can be removed by script, the **Docs Content Validation** system will probably stop any such PR in its tracks (unless admin merges anyway).

File: aspnetcore/host-and-deploy/health-checks.md [hunk 1]
```diff
@@ -633,6 +635,45 @@ The `IHealthCheckPublisher` interface has a single method:
...
+* `Period` &ndash; The period of `IHealthCheckPublisher` execution. The default value is 30 seconds.
+* `Predicate` &ndash; If `Predicate` is `null` (default), the health check publisher service runs all registered health checks. To run a subset of health checks, provide a function that filters the set of checks. The predicate is evaluated each period.
+* `Timeout` &ndash; The timeout for executing the health checks for all `IHealthCheckPublisher`instances. Use <xref:System.Threading.Timeout.InfiniteTimeSpan> to execute without a timeout. The default value is 30 seconds.
+
+::: moniker range="= aspnetcore-2.2"
+
+> [!WARNING]
+> Due to an issue in the ASP.NET Core 2.2 release, setting `Period` isn't honored by the `IHealthCheckPublisher` implementation and inadvertently sets the value of `Delay`. This issue will be fixed in ASP.NET Core 3.0. For more information, see [HealthCheckPublisherOptions.Period sets the value of .Delay](https://github.com/aspnet/Extensions/issues/1041).
```
Comment (Rick-Anderson) [inline aspnetcore/host-and-deploy/health-checks.md]: In the ASP.NET Core 2.2 release, setting `Period` isn't honored by the `IHealthCheckPublisher` implementation; it sets the value of `Delay`. This issue will be fixed in ASP.NET Core 3.0. For more information, see [HealthCheckPublisherOptions.Period sets the value of .Delay](https://github.com/aspnet/Extensions/issues/1041).

**Comment by guardrex** [comment]
@rynowak Keep an :eye: on my namespaces. I'm using `Microsoft.AspNetCore.Hosting` in *Program.cs* but `Microsoft.Extensions.Hosting` elsewhere ... scatterbrained. I need `IWebHost` and `WebHostBuilder` tho in *Program.cs*. What's the recommended approach?

**Comment by rynowak** [comment]
We're making some changes in 3.0 to try and retire or de-emphasize from the of the web host builder. What version are these samples supposed to target? If this is for 2.2 then it seems like a good bet to have a using for `Microsoft.AspNetCore.Hosting` and use fully-qualified names for `Microsoft.Extensions.Hosting` where needed. 

/cc @Tratcher in case he has the thoughts

**Comment by Tratcher** [comment]
For 2.2 use AspNetCore.Hosting. We know the namespace conflict is a pain and we plan to fix it in 3.0 https://github.com/aspnet/Extensions/issues/966

**Comment by guardrex** [comment]
I wasn't using `IHostingEnvironment` in one spot, so I was able to drop that. I converted others to `Microsoft.AspNetCore.Hosting`.

I think we're ok now. Anything else to do here?

**Comment by guardrex** [comment]
@Tratcher @rynowak Made the namespace updates on the last commit.

The only usages of `Microsoft.Extensions.Hosting` are for `IHostingService` usages on ...

```csharp
services.TryAddEnumerable(
    ServiceDescriptor.Singleton(typeof(IHostedService), 
        typeof(HealthCheckPublisherOptions).Assembly
            .GetType(HealthCheckServiceAssembly)));
```

... which is a temporary workaround for 2.2 (not needed for 3.0) ... and ...

```csharp
public class StartupHostedService : IHostedService, IDisposable
```

If those two usages are ok, we're good.

**Comment by guardrex** [comment]
We're going to go ahead ..... 🏃🏃🏃🏃🏃🏃 .... trying to clear out our PRs in advance of major 3.0 work.

**Comment by Tratcher** [comment]
That's fine.

**Pull request merged**